### PR TITLE
Add np.vsplit, np.hsplit, and np.dsplit

### PR DESCRIPTION
**Pull request by KrisMinchev**
Adding implementations of `np.vsplit`,  `np.hsplit`, and `np.dsplit` to Numba, as requested in issue https://github.com/numba/numba/issues/4074. Note that these are special cases of the `np.split` function.

### CODE REVIEW

File: numba/np/arrayobj.py [hunk 1]
```diff
@@ -6102,6 +6102,48 @@ def impl(ary, indices_or_sections, axis=0):
...
+@overload(np.hsplit)
+def numpy_hsplit(ary, indices_or_sections):
+    if not isinstance(ary, types.Array):
+        msg = 'The argument "ary" must be an array'
+        raise errors.TypingError(msg)
+
+    if not isinstance(indices_or_sections, (types.Integer, types.Array)):
+        msg = ('The argument "indices_or_sections" '
+               'must be an integer or an array')
+        raise errors.TypingError(msg)
```
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:6134]: Although NumPy says the second argument is an array, the following is also valid:

```python
import numpy as np
from numba import njit

@njit
def foo(a):
    return np.hsplit(a, [3, 6])

x = np.arange(16).reshape(4, 4)
# print(foo(x))
print(foo.py_func(x))
```

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @KrisMinchev, thanks for working on this. I've found a case where your code fails: NumPy docs says the second argument to `vsplit`/`hsplit` must be an integer or 1D-array, but any iterable is accepted.

File: numba/np/arrayobj.py [hunk 2]
```diff
@@ -6102,6 +6102,71 @@ def impl(ary, indices_or_sections, axis=0):
...
+@overload(np.vsplit)
+def numpy_vsplit(ary, indices_or_sections):
+    if not isinstance(ary, types.Array):
+        msg = 'The argument "ary" must be an array'
+        raise errors.TypingError(msg)
+
+    if not isinstance(indices_or_sections, (types.Integer, types.Array,
+                                            types.List, types.UniTuple)):
+        msg = ('The argument "indices_or_sections" '
+               'must be an integer, an array, a list, or a tuple')
```
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:6114]:
```suggestion
        msg = ('The argument "indices_or_sections" must be int or 1d-array')
```

File: numba/np/arrayobj.py [hunk 3]
```diff
@@ -6102,6 +6102,71 @@ def impl(ary, indices_or_sections, axis=0):
...
+@overload(np.hsplit)
+def numpy_hsplit(ary, indices_or_sections):
+    if not isinstance(ary, types.Array):
+        msg = 'The argument "ary" must be an array'
+        raise errors.TypingError(msg)
+
+    if not isinstance(indices_or_sections, (types.Integer, types.Array,
+                                            types.List, types.UniTuple)):
+        msg = ('The argument "indices_or_sections" '
+               'must be an integer, an array, a list, or a tuple')
```
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:6135]:
```suggestion
        msg = ('The argument "indices_or_sections" must be int or 1d-array')
```

File: numba/np/arrayobj.py [hunk 4]
```diff
@@ -6102,6 +6102,71 @@ def impl(ary, indices_or_sections, axis=0):
...
+@overload(np.dsplit)
+def numpy_dsplit(ary, indices_or_sections):
+    if not isinstance(ary, types.Array):
+        msg = 'The argument "ary" must be an array'
+        raise errors.TypingError(msg)
+
+    if not isinstance(indices_or_sections, (types.Integer, types.Array,
+                                            types.List, types.UniTuple)):
+        msg = ('The argument "indices_or_sections" '
+               'must be an integer, an array, a list, or a tuple')
```
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:6158]:
```suggestion
        msg = ('The argument "indices_or_sections" must be int or 1d-array')
```

File: numba/tests/test_np_functions.py [hunk 5]
```diff
@@ -3181,6 +3193,155 @@ def test_split_basic(self):
...
+        # only do simple test on vsplit
+        def inputs():
+            yield np.array([[1, 2, 3, 4], [1, 2, 3, 4]]), 2
+            yield np.array([[1., 2., 3., 4.], [1., 2., 3., 4.]]), 2
+            yield np.arange(16.0).reshape(4, 4), 2
+            yield np.arange(16.0).reshape(4, 4), np.array([3, 6])
+            yield np.arange(16.0).reshape(4, 4), [3, 6]
+            yield np.arange(16.0).reshape(4, 4), (3, 6)
+            yield np.arange(8.0).reshape(2, 2, 2), 2
+
```
Comment (guilhermeleobas) [inline numba/tests/test_np_functions.py:3207]: Are these tests from the NumPy test suite? If yes, add a link referencing the NumPy GitHub repo

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
One last thing and it should be good to merge

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
Hi @KrisMinchev, thanks again. I have a few more request changes and PR should be good to merge

File: numba/tests/test_np_functions.py [hunk 6]
```diff
@@ -3181,6 +3193,162 @@ def test_split_basic(self):
...
+        pyfunc = vsplit
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for a, ind_or_sec in inputs():
+            self.assertPreciseEqual(pyfunc(a, ind_or_sec),
+                                    cfunc(a, ind_or_sec))
+
+    def test_vsplit_exception(self):
```
[the remark this answers is not in the document]
Comment (KrisMinchev) [reply numba/tests/test_np_functions.py:3218]:
```suggestion
    def test_vhdsplit_exception(self):
        # Single test method for vsplit, hsplit and dsplit
        for (f, mindim, name) in [(vsplit, 2, "vsplit"),
                                  (hsplit, 1, "hsplit"),
                                  (dsplit, 3, "dsplit")]:
            cfunc = jit(nopython=True)(f)

            self.disable_leak_check()

            with self.assertRaises(TypingError) as raises:
                cfunc(1, 2)
            self.assertIn('The argument "ary" must be an array',
                        str(raises.exception))

            with self.assertRaises(TypingError) as raises:
                cfunc("abc", 2)
            self.assertIn('The argument "ary" must be an array',
                        str(raises.exception))

            with self.assertRaises(TypingError) as raises:
                cfunc(np.array([[1, 2, 3, 4], [1, 2, 3, 4]]), "abc")
            self.assertIn(('The argument "indices_or_sections" must be int or '
                        '1d-array'),
                        str(raises.exception))

            with self.assertRaises(ValueError) as raises:
                cfunc(np.array(1), 2)
            self.assertIn(name + ' only works on arrays of ' + str(mindim) +
                          ' or more dimensions',
                        str(raises.exception))
```

[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/tests/test_np_functions.py:3220]: Great! That should be enough

**Comment by KrisMinchev** [comment]
Hi @guilhermeleobas, thank you for your requests. I have addressed them in my last commits.

File: numba/tests/test_np_functions.py [hunk 7]
```diff
@@ -3181,6 +3193,109 @@ def test_split_basic(self):
...
+            yield np.array([1, 2, 3, 4]), 2
+            yield np.array([1., 2., 3., 4.]), 2
+            # test_2D_array
+            yield np.array([[1, 2, 3, 4], [1, 2, 3, 4]]), 2
+            yield np.array([[1., 2., 3., 4.], [1., 2., 3., 4.]]), 2
+            yield np.arange(16.0).reshape(4, 4), 2
+            yield np.arange(16.0).reshape(4, 4), np.array([3, 6])
+            yield np.arange(16.0).reshape(4, 4), [3, 6]
+            yield np.arange(16.0).reshape(4, 4), (3, 6)
+            yield np.arange(8.0).reshape(2, 2, 2), 2
```
Comment (guilhermeleobas) [inline numba/tests/test_np_functions.py:3238]: Tests from hsplit/vsplit/dsplit are the same. Merge the functions into a single one

File: numba/tests/test_np_functions.py [hunk 8]
```diff
@@ -3181,6 +3193,109 @@ def test_split_basic(self):
...
+            self.assertPreciseEqual(pyfunc(a, ind_or_sec),
+                                    cfunc(a, ind_or_sec))
+
+    def test_hsplit_basic(self):
+        # split and array_split have more comprehensive tests of splitting.
+        # only do simple test on hsplit
+        def inputs():
+            # Taken from https://github.com/numpy/numpy/blob/...
+            # 11a2ea96a7ac923c2283d487bffa7fce8924cbd7/numpy/lib/tests/...
+            # test_shape_base.py#L553-L564
```
Comment (guilhermeleobas) [inline numba/tests/test_np_functions.py:3226]: Keep the URL in a single line

File: numba/tests/test_np_functions.py [hunk 9]
```diff
@@ -3181,6 +3193,109 @@ def test_split_basic(self):
...
+            self.assertPreciseEqual(pyfunc(a, ind_or_sec),
+                                    cfunc(a, ind_or_sec))
+
+    def test_dsplit_basic(self):
+        # split and array_split have more comprehensive tests of splitting.
+        # only do simple test on dsplit
+        def inputs():
+            # Taken from https://github.com/numpy/numpy/blob/...
+            # 11a2ea96a7ac923c2283d487bffa7fce8924cbd7/numpy/lib/tests/...
+            # test_shape_base.py#L616-L624
```
Comment (guilhermeleobas) [inline numba/tests/test_np_functions.py:3253]: Keep the URL in a single line

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]

**Comment by gmarkall** [comment]
(Note: the build fail is just a flake8 issue)

**Review by guilhermeleobas — APPROVED** [review]
Thanks @KrisMinchev

**Pull request merged**